lib/elixir_sense/core/compiler.ex

```diff
@@ -0,0 +1,4696 @@
+
+        [] ->
+          # deprecated @attribute()
+          {false, {nil, state, env}}
+
+        [_] ->
+          # @attribute(arg)
+          # elixir validates env.function is nil
+          # elixir forbids behavior name
+          {true, expand_args(args, state, env)}
+
+        args ->
+          # elixir raises "invalid @ call #{inspect(args)}"
+          {e_args, state, env} = expand_args(args, state, env)
+          {true, {[hd(e_args)], state, env}}
+      end
+
+    inferred_type =
+      case e_args do
+        nil -> nil
+        [arg] -> TypeInference.type_of(arg, env.context)
+      end
+
+    state =
+      state
+      |> add_attribute(env, name, meta, e_args, inferred_type, is_definition)
+      |> add_current_env_to_line(meta, env)
+
+    {{:@, meta, [{name, name_meta, e_args}]}, state, env}
+  end
+
+  defp expand_macro(
+         meta,
+         Kernel,
+         :defoverridable,
+         [arg],
+         _callback,
+         state,
+         env = %{module: module}
+       )
+       when module != nil do
+    {arg, state, env} = expand(arg, state, env)
+
+    case arg do
+      keyword when is_list(keyword) ->
+        {nil, make_overridable(state, env, keyword, meta[:context]), env}
+
+      behaviour_module when is_atom(behaviour_module) ->
+        if Code.ensure_loaded?(behaviour_module) and
+             function_exported?(behaviour_module, :behaviour_info, 1) do
+          keyword =
+            behaviour_module.behaviour_info(:callbacks)
+            |> Enum.map(&Introspection.drop_macro_prefix/1)
+
+          {nil, make_overridable(state, env, keyword, meta[:context]), env}
+        else
+          {nil, state, env}
+        end
+
+      _ ->
+        {nil, state, env}
+    end
+  end
+
+  defp expand_macro(
+         meta,
+         Kernel,
+         type,
+         [fields],
+         _callback,
+         state,
+         env = %{module: module}
+       )
+       when type in [:defstruct, :defexception] and module != nil do
+    if Map.has_key?(state.structs, module) do
+      raise ArgumentError,
+            "defstruct has already been called for " <>
+              "#{inspect(module)}, defstruct can only be called once per module"
+    end
+
+    {fields, state, env} = expand(fields, state, env)
+
+    fields =
+      case fields do
+        fs when is_list(fs) ->
+          fs
+
+        _other ->
+          # elixir raises ArgumentError here
+          []
+      end
+
+    fields =
+      fields
+      |> Enum.filter(fn
+        field when is_atom(field) -> true
+        {field, _} when is_atom(field) -> true
+        _ -> false
+      end)
+      |> Enum.map(fn
+        field when is_atom(field) -> {field, nil}
+        {field, value} when is_atom(field) -> {field, value}
+      end)
+
+    state =
+      state
+      |> add_struct_or_exception(env, type, fields, extract_range(meta))
+
+    {{type, meta, [fields]}, state, env}
+  end
+
+  defp expand_macro(
+         meta,
+         Record,
```

Comment:
I think it is fine to special case Kernel, but it would be nice to make `Record` and `Protocol` functions work without special casing entries here.

Reply:
It's hard to do that without eval though. In case of `Record` I plan to extract field info similarly to what I do with `defstruct`